### ISSUE TITLE
Collection Change Handlers: AppKit-Compatible Indexes

### ISSUE THREAD

**Issue by bdkjones** [realm/realm-swift#7980]
### Problem

After filing #339 as a bug and discovering the design was intentional, I'm filing this as a feature request.

In collection change handlers, the `modifications` array of indexes references the *previous* collection. That seems to be appropriate for `UITableView` on iOS. However, it does not work for `NSTableView` on macOS.

On macOS, the indexes for modified objects are useless if any insertions and/or deletions have occurred in the same change notification. AppKit developers have to force-reload the entire table because there is no reliable way to translate the "old" modified-object indexes that Realm provides to the CURRENT indexes of those objects, so we don't know which rows of a table need to be reloaded to display the modified data.

### Solution

The easiest solution I see is to provide *two* `modifications` arrays: one referencing the previous collection version and one referencing the latest collection version:

```
switch changes
{
    case .initial:
        break
            
    case .update(_, let deletions, let insertions, let modifications, let currentModifications):
        // "currentModifications" is [Int] referencing the current state of the Realm Collection.
```

### Other Options:

1. Have `modifications` reference the *current* state of the collection if the application is linked against `AppKit` instead of `UIKit`
3. Allow the developer to specify which kind of `modifications` he wants (old collection state or current collection state) in the `.observe()` call. (The default would be as-is behavior for backward compatibility.)

Either way, it would be useful if Mac apps didn't have to resort to a full-reload of `NSTableView` when handling changesets. 

### Just Use SwifUI!

For Mac apps, which frequently have much more complex UI, SwiftUI simply isn't ready. It will be many years before that framework is ready to lift the heavy UI loads that AppKit handles (32-column, 8,000 row tables, etc.).

### How important is this improvement for you?

Would be a major improvement

**Comment by bdkjones** [comment]
Nothing on this proposal? Realm *does* list macOS as an official platform, but this API is essentially UIKit-only. It would be nice if Realm were a proper fit on all the platforms it supports.

![Screenshot 2024-05-01 at 01 17 49](https://github.com/realm/realm-swift/assets/926566/1a28dd8b-9949-479e-ab22-4d36d873701c)

**Comment by bdkjones** [comment]
The problem is more extensive than just `modifications && (insertions || deletions)`. The problem can manifest anytime there are BOTH `insertions` and `deletions` in a changeset. 

This happens, for example, if an `NSTableView` is sorted by a column, `foo`, and the user changes the value of that column in one row such that the sort order is affected. A changeset will happen with `1 deletion` and `1 insertion`, but again, once the deletions are handled, the rows for insertions are now wrong—they reference the OLD state of the table's data. 

## Bottom Line

If you're a Mac developer and you use `NSTableView` with a Realm `Results` object, you can't follow the docs. If either of these things is true, you must force-reload the entire table:

1) `modifications` is non-zero and EITHER `deletions` or `insertions` is non-zero.

2) both `deletions` and `insertions` are non-zero, regardless of the value of `modifications`.

**Comment by bdkjones** [comment]
@nirinchev wouldn't it be non-breaking to simply add an option to the collection change subscription that controls how the indexes are calculated? The default would be the current behavior (indexes relative to the "old" collection, suitable for UIKit's -batchUpdates()) and AppKit-based developers could specify an option that would give them the indexes they need. That wouldn't change anything for existing implementations?

**Comment by nirinchev** [comment]
Yes, that's one approach we're considering but I'd like to see if there's anything smart we could do that exposes both sets of indexes, like we do in other SDKs.
